rust/src/api.rs

```diff
@@ -139,29 +117,132 @@ pub fn decode_invoice(invoice: String) -> Result<LightningInvoice> {
     })
 }
 
+#[derive(Clone)]
+pub enum Event {
+    Init(String),
+    Ready,
+    WalletInfo(WalletInfo),
+    Offer(Option<Offer>),
+}
+
+#[derive(Clone)]
+pub struct WalletInfo {
+    pub balance: Balance,
+    pub bitcoin_history: Vec<BitcoinTxHistoryItem>,
+    pub lightning_history: Vec<LightningTransaction>,
+}
+
+impl WalletInfo {
+    fn new() -> WalletInfo {
+        WalletInfo {
+            balance: Balance {
+                on_chain: OnChain {
+                    trusted_pending: 0,
+                    untrusted_pending: 0,
+                    confirmed: 0,
+                },
+                off_chain: OffChain {
+                    available: 0,
+                    pending_close: 0,
+                },
+            },
+            bitcoin_history: vec![],
+            lightning_history: vec![],
+        }
+    }
+
+    async fn build_wallet_info() -> Result<WalletInfo> {
+        let balance = wallet::get_balance()?;
+        let bitcoin_history = get_bitcoin_tx_history().await?;
+        let lightning_history = wallet::get_lightning_history().await?;
+        Ok(WalletInfo {
+            balance,
+            bitcoin_history,
+            lightning_history,
+        })
+    }
+}
+
+#[tokio::main(flavor = "current_thread")]
+pub async fn refresh_wallet_info() -> Result<WalletInfo> {
+    WalletInfo::build_wallet_info().await
+}
+
 #[tokio::main(flavor = "current_thread")]
-pub async fn run_ldk() -> Result<()> {
+pub async fn run(stream: StreamSink<Event>, app_dir: String) -> Result<()> {
+    let network = config::network();
+    anyhow::ensure!(!app_dir.is_empty(), "app_dir must not be empty");
+    stream.add(Event::Init(format!("Initialising {network} wallet")));
+    wallet::init_wallet(Path::new(app_dir.as_str()))?;
```

Comment:
I'm afraid that with one single `run()` method we might have trouble controlling the invariants established  in #558 
(not impossible, but it's hidden deeper)

Reply:
can you elaborate on that?